swagger.yaml

```diff
@@ -3786,6 +3786,88 @@ paths:
           schema:
             $ref: '#/definitions/error'
 
+  /v2/clusters/{cluster_id}/ui-settings:
```

Comment:
I moved this comment to here so the entire conversation will be in single thread.
I think that there is something that I might missing here. Why do we need that? Only for custom manifests? If so so why not just add a boolean to the cluster that indicates that the user selected the custom manifests or there are custom manifests.
We can use this indicator in different places to know if there are custom manifests of not without calling to /v2/clusters/{cluster_id}/manifests (e.g validations, feature-usage and more)
Further more this can cause incompatibility issues when the user uses API + UI (some flows can be done only when using the API)

Reply:
@avishayt @gamli75 @paul-maidment @jkilzi

Reply:
It is not going to be only for custom manifests. The API is for persisting UI-only state.

Reply:
Also, the intention of this boolean is not as you describe it

This setting is **UI Specific**, meaning that it is only to be used to determine the state of a UI component (in this case the "tab" that determines whether or not you will be able to add custom manifests to your installation.)

A cluster with this UI preference set is not guaranteed to have custom manifests uploaded to it. A connection to S3 would still be required to determine whether or not there are manifests available.

I understand the confusion but the intent and scope of this field does not go beyond the UI.

In discussion with @filanov and @jkilzi it was agreed that the UI team would in fact like a field to hold generic UI values, owned by the UI team for which they would take responsibility.

Reply:
Why does it need to be a new api? why not just make it as part of cluster create and update?

Reply:
The intent of this field is to make sure that it is for UI specific uses. 

Would it be acceptable to export this field (which has a UI specific purpose) when giving API responses via the UI. It is my understanding that this would be necessary if we take the approach suggested by @filanov 

@avishayt do you have any thoughts on this?

Reply:
I think it is more of a separation of concerns. Why does the Cluster as a stand-alone entity need an attribute called ui_settings?

Reply:
`I think it is more of a separation of concerns. Why does the Cluster as a stand-alone entity need an attribute called ui_settings?` @jkilzi - I agree

Reply:
On the one hand, I agree that separate is nicer in this case.  However, will the UI need to "atomically" update both the regular cluster properties and the UI settings?  If so, a cluster property would help.

Reply:
@jkilzi is this the case, does the UI need to atomically update cluster state and UI settings?

Reply:
Elior just to make it clear this field is used by the ui to store specific status for the wizard, for example if user is planning to add custom manifests or static network or any other field in the future, instead of having a new boolean each time we prefer to have a field that the ui will manage and store the data there

Reply:
> Elior just to make it clear this field is used by the ui to store specific status for the wizard,

I know, but while working on https://github.com/openshift/assisted-service/pull/5364 I needed this exact property. I think that 

> is not relevant we are not support it

Currently, but why to create this separation now? I think that the objects that we currently have (e.g. cluster, infraEnv.. ) supposed to reflect the state. To my opinion when we add this kind of API we are allowing to hold another state on the same objects, that for the long run I believe will cause us more headaches.
For example what about feature usages? If the user mark the `Custom Manifests` checkbox, how can we tell that the user did that? 

> instead of having a new boolean each time we prefer to have a field that the ui will manage and store the data there

I don't think that we will need to add a boolean each time, but in this case I think we should (from the reasons I mentioned above). If I encountered this need only few days ago I'm sure that others will need i n the future.

Reply:
"I know, but while working on https://github.com/openshift/assisted-service/pull/5364 I needed this exact property. I think that"
It appears that the purpose of that PR is to block installation if there are no manifests uploaded, the subtle distinction here is that this flag does not guarantee that user manifests are available as it exists for the sole purpose of storing a user selected "checkbox" to indicate that they wish to add custom manifests.

It is still entirely possible that the user will make this selection and then fail to add manifests. So this boolean does not serve the purpose you describe. (not without very strict validation/enforcement prior to starting installation at least.)

The purpose of this checkbox is completely for UI use, it has nothing to do with the cluster domain.


"Currently, but why to create this separation now? " -
The way I view this is that we place the UISettings "JSON blob" object inside the cluster database table only for convenience and to save the expense of creating a completely independent datastore, for this task it's overkill.

I see this as an implementation detail rather than something that belongs to the cluster. The API call encapsulates this neatly and allows for the furture possibility of improving this isolatin by moving the underlying data to a more appropriate structure. 

To tie this data structure to the cluster by exposing the field in the API is tight coupling, generally discouraged as good design practice.

Moreover it exists (at the request of the UI team) for the purpose of providing a simple way in which they can store "UI settings" for a specific "workflow" in a cluster.

I personally think that this PR strikes the right balance of separation of concerns and capability to keep most stakeholders happy.

I do hear the concern about having a field that would negate the need to connect to S3 to simply validate that the user has uploaded custom manifests. My thoughts are that this should be addressed elsewhere rather than in this PR, for which the focus is strictly UI.

On that topic, I have my own apprehensions about why we even use S3 to store manifests instead of a database, cases like the one identified by @eliorerz would be greatly simplified, but realise that a migration would be costly and require a lot of work.

Reply:
The UI should only use this field for UI state that doesn't affect the cluster/host/infraenv resources.  In the example of custom manifests, the UI isn't saving the state of "the user added custom manifests" but rather "the user clicked that they want to add custom manifests so we should show that step in the wizard".  Another example might be "the user didn't choose platform integration in the first screen, so we'll allow them to choose UMN/CMN in the network screen".  This is more straight-forward to save in a special UI field than adding random fields to the cluster resource.

Reply:
Agree with what Avishay said, in addition i don't think that https://github.com/openshift/assisted-service/pull/5364 is needed if we have this api, the ui will not create an empty manifest and we don't care for the logic in the wizard for this case.

Reply:
We still need #5364 - it validates that a custom manifest was provided for the OCI platform

Reply:
For the record, the UI doesn't need to update this endpoint and the cluster endpoint atomically.